test/helpers/fuzzing.js

```diff
@@ -0,0 +1,242 @@
+      expectedWrites[singleWrite.key] = singleWrite.values.filter(val => !!val)
+    }
+  }
+  var allKeys = Object.keys(expectedWrites)
+
+  t.test(['validating after', processedBatches.length, 'replications'].join(' '), function (t) {
+    t.plan(allKeys.length + 1)
+
+    var readStream = db.createReadStream('/')
+    readStream.on('end', function () {
+      var keys = Object.keys(expectedWrites)
+      t.same(keys.length, 0, 'missing keys: ' + keys)
+      if (keys.length === 0) return cb()
+      return cb(new Error(`missing keys: ${keys}`))
+    })
+    readStream.on('error', cb)
+    readStream.on('data', function (nodes) {
+      if (!nodes) return
+      var key = nodes[0].key
+      var values = nodes.map(node => node.value)
+      t.same(values, expectedWrites[key])
+      delete expectedWrites[key]
+    })
+  })
+}
+
+function generateFailingTest (name, dbCount, writesPerReplication, writeOps, cb) {
+  writeOps.push(err => {
+    t.error(err)
+    t.end()
+  }).toString()
+  fs.writeFile(p.join(__dirname, '..', name + '.js'), prettier.format(`
+    var tape = require('tape')  
+
+    var run = require('./helpers/run')
+    var put = require('./helpers/put')
+    var create = require('./helpers/create')
+
+    var validate = require('./helpers/fuzzing').validate
+
+    var writesPerReplication = ${JSON.stringify(writesPerReplication)}
+
+    tape('autogenerated failing fuzz test', function (t) {
+      create.many(${dbCount}, function (err, dbs, replicateByIndex) {
+        t.error(err)
+        run(${writeOps.map(op => op.toString())})
+      })
+    })`, { singleQuote: true, semi: false }, { encoding: 'utf-8' }, cb))
+}
+
+var testNum = 0
+
+function fuzzRunner (t, opts, cb) {
+  opts = defaultOpts(opts)
+
+  // Used for keeping track of failing test cases.
+  testNum++
+
+  var writesPerReplication = generateData(opts)
+
+  makeDatabases(opts, function (dbs, replicateByIndex) {
+    var ops = []
+    for (var i = 0; i < writesPerReplication.length; i++) {
+      var batch = writesPerReplication[i]
+      var batchOps = []
+      for (var j = 0; j < batch.length; j++) {
+        var singleWrite = batch[j]
+        for (var z = 0; z < singleWrite.values.length; z++) {
+          var value = singleWrite.values[z]
+          if (!value) continue
+          batchOps.push((function (z, k, v) {
+            return eval(`(cb => {
+              put(dbs[${z}], [{
+                key: '${k}',
+                value: '${v}'
+              }], cb)
+            })`)
+          })(z, singleWrite.key, value))
+        }
+      }
+      ops.push(batchOps)
+      // Intersperse replication/validation/failing-test generation between write batches.
+      ops.push([
+        // Currently replicating between all databases at every replication point.
+        cb => replicateByIndex(cb),
+        (function (i) {
+          return eval(`(cb => validate(t, dbs[0], writesPerReplication.slice(0, ${i + 1}), cb))`)
+        })(i),
```

Comment:
dont think you need the fn wrapper

Reply:
Ya removed the two cases where I was wrapping. It isn't necessary w/ the eval.